src/restic/node.go

```diff
@@ -191,7 +191,19 @@ func (node Node) createDirAt(path string) error {
 	return nil
 }
 
-func (node Node) createFileAt(path string, repo Repository) error {
+func (node Node) createFileAt(path string, repo Repository, idx *HardlinkIndex) error {
+	if node.Links > 1 {
+		if !idx.ExistsLink(node.Inode, node.Device) {
+			idx.AddLink(node.Inode, node.Device, path)
```

Comment:
I think the node should only be added to the hardlink index when the restore was successful, so you need to move this to the end of the function. Otherwise restoring a file may fail, and with the current code all hardlinked versions of that file will also not be restored because the hardlink index contains a filename that may not exist (or even point to different content).